exchanges/kraken/kraken_websocket.go

```diff
@@ -1105,211 +924,365 @@ func trim(s string) string {
+		return fmt.Errorf("error parsing WS status: %w from message: %s", err, resp)
+	} else if status != exp {
+		return fmt.Errorf("wrong WS status: %s; expected: %s from message %s", exp, op, resp)
+	}
 
-		var id int64
-		if common.StringDataContains(authenticatedChannels, channelsToUnsubscribe[x].Channel) {
-			id = k.Websocket.AuthConn.GenerateMessageID(false)
-		} else {
-			id = k.Websocket.Conn.GenerateMessageID(false)
-		}
+	return nil
+}
 
-		unsub := WebsocketSubscriptionEventRequest{
-			Event: krakenWsUnsubscribe,
-			Pairs: []string{channelsToUnsubscribe[x].Pairs[0].String()},
-			Subscription: WebsocketSubscriptionData{
-				Name:  channelsToUnsubscribe[x].Channel,
-				Depth: depth,
-			},
-			RequestID: id,
-		}
-		if common.StringDataContains(authenticatedChannels, channelsToUnsubscribe[x].Channel) {
-			unsub.Subscription.Token = authToken
+// getRespErr takes a json response string and looks for an error event type
+// If found it returns the errorMessage
+// It might log parsing errors about the nature of the error
+// If the error message is not defined it will return a wrapped errUnknownError
+func (k *Kraken) getRespErr(resp []byte) error {
+	event, err := jsonparser.GetUnsafeString(resp, "event")
+	switch {
+	case err != nil:
+		return fmt.Errorf("error parsing WS event: %w from message: %s", err, resp)
+	case event != "error":
+		status, _ := jsonparser.GetUnsafeString(resp, "status") // Error is really irrellevant here
+		if status != "error" {
+			return nil
 		}
-		unsub.Channels = append(unsub.Channels, channelsToUnsubscribe[x])
-		unsubs = append(unsubs, unsub)
 	}
 
-	var errs error
-	for i := range unsubs {
-		var err error
-		if common.StringDataContains(authenticatedChannels, unsubs[i].Subscription.Name) {
-			_, err = k.Websocket.AuthConn.SendMessageReturnResponse(unsubs[i].RequestID, unsubs[i])
-		} else {
-			_, err = k.Websocket.Conn.SendMessageReturnResponse(unsubs[i].RequestID, unsubs[i])
+	var msg string
+	if msg, err = jsonparser.GetString(resp, "errorMessage"); err != nil {
+		log.Errorf(log.ExchangeSys, "%s error parsing WS errorMessage: %s from message: %s", k.Name, err, resp)
+		return fmt.Errorf("error status did not contain errorMessage: %s", resp)
+	}
+	return errors.New(msg)
+}
+
+// wsProcessSubStatus handles creating or removing Subscriptions as soon as we receive a message
+// It's job is to ensure that subscription state is kept correct sequentially between WS messages
+// If this responsibility was moved to Subscribe then we would have a race due to the channel connecting IncomingWithData
+// All errors are ignored; We only care about success
+func (k *Kraken) wsProcessSubStatus(resp []byte) {
+	pName, err := jsonparser.GetUnsafeString(resp, "pair")
+	if err != nil {
+		return
+	}
+	pair, err := currency.NewPairFromString(pName)
+	if err != nil {
+		return
+	}
+	channelName, err := jsonparser.GetUnsafeString(resp, "channelName")
```

Comment:
You're saying to not shadow a func name with a variable name, right?
This policy would have further reaching implications both in this PR, and in our source code in general,
so I just want to be sure we're on the same page.

If the variable was used as a function it would throw a compile time eror.

There doesn't seem to be a linter rule about this.
Is your IDE complaining about it?

Reply:
Fixed 8b5c3774

Reply:
It's more along the lines of "must you?" 😄 You're introducing a function in this PR called channelName, and then using a var name which has a different purpose. It's about asking you whether any other names for either the variables or the function can be made more appropriate